logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
             Map<Reference, Set<Label>> mapping = new HashMap<>();
-            traversalEng.iterator(conjunction.traversal()).forEachRemaining(
+            traversalEng.iterator(traversal).forEachRemaining(
                     result -> result.forEach((ref, vertex) -> {
                         mapping.putIfAbsent(ref, new HashSet<>());
-                        mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
+                        assert vertex.isType();
+                        if (!(vertex.asType().isAbstract() && traversalConstructor.getVariable(ref).isThing()))
+                            mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
                     })
             );
             return mapping;
         });
     }
 
-    private Type getType(Label label) {
-        if (label.scope().isPresent()) {
-            assert conceptMgr.getRelationType(label.scope().get()) != null;
-            return conceptMgr.getRelationType(label.scope().get()).getRelates(label.name());
-        } else {
-            return conceptMgr.getThingType(label.name());
-        }
-    }
+    private static class TraversalConstructor {
 
-    private static class ConstraintMapper {
-        private final Resolvers resolvers;
-        private final HashMap<TypeVariable, Set<TypeVariable>> neighbours;
-        private final TypeVariable rootAttributeType;
-        private final TypeVariable rootRelationType;
-        private final TypeVariable rootRoleType;
         private final Conjunction conjunction;
+        private final Map<Reference, Variable> variableRegister;
+        private final Map<Identifier, TypeVariable> resolverRegister;
+        private final Traversal resolverTraversal;
+        private int sysVarCounter;
         private final ConceptManager conceptMgr;
+        private final Map<Identifier, Set<ValueType>> valueTypeRegister;
 
-        ConstraintMapper(Conjunction conjunction, ConceptManager conceptMgr) {
-            this.conceptMgr = conceptMgr;
-            this.resolvers = new Resolvers();
-            this.neighbours = new HashMap<>();
+        TraversalConstructor(Conjunction conjunction, ConceptManager conceptMgr) {
             this.conjunction = conjunction;
-            this.rootAttributeType = retrieveTypeFromLabel(Label.of(ATTRIBUTE.toString()));
-            this.rootRelationType = retrieveTypeFromLabel(Label.of(RELATION.toString()));
-            this.rootRoleType = retrieveTypeFromLabel(Label.of(ROLE.toString(), RELATION.toString()));
+            this.conceptMgr = conceptMgr;
+            this.resolverTraversal = new Traversal();
+            this.variableRegister = new HashMap<>();
+            this.resolverRegister = new HashMap<>();
+            this.sysVarCounter = 0;
+            this.valueTypeRegister = new HashMap<>();
             conjunction.variables().forEach(this::convert);
```

Comment:
This is the main loop of the code. We take each variable and convert its constraints into edges on a `Traversal`.